### ISSUE TITLE
Discussion to add autocmd-event for setloclist/setqflist

### ISSUE THREAD

**Issue by weilbith** [neovim/neovim#13099]
### Actual behaviour
There is only an event emitted before and after a command as run that Vim considers to be relevant for the quickfix or location list (e.g. `:grep`). But the lists can be set by basically everyone, causing no event the user could listen to.

### Expected behaviour
An event gets thrown when ever `setqflist` or `setloclist` got called.

---

Sorry if this was already discussed once. I was not able to find such conversation. 
I know there should not be an event for every rare use-case. But in fact this event(s) do already exist, just that they do not apply not for everything. Why emitting an event for specific commands only rather than make it more generic?
To finally make an example where this could be useful would be a status segment that highlights if and how much location lists entry exist. So there is no need to automatically open the location list (if such feature is provided), but the user can have a visual feedback if any asynchronous tool is adding location or quickfix list entries.
Thanks for any input!

**Comment by emmanueltouzery** [comment]
another example of a use-case is the https://github.com/dhruvasagar/vim-markify plugin, which displays signs for quickfix locations and uses `QuickFixCmdPost` to listen to quickfix changes, but that even is not triggered for manual `setqflist` changes.

**Comment by justinmk** [comment]
See also https://github.com/vim/vim/discussions/16181

**Comment by telemachus** [comment]
@hfn92 I think you are almost there, but—if the context is `setqflist` or `setloclist`—you need to make a [User autocommand](https://neovim.io/doc/user/autocmd.html#User) and then manually fire the autocommand that you create. Here's my solution for something I wrote for myself. (This may not be the most elegant solution, but it works.)

First, in an initialization function, I create the User autocommand:

```lua
 local refinery_augroup =
        vim.api.nvim_create_augroup("RefineryAugroup", { clear = true })
    -- Create a no-op autocommand to fire after filling QuickFix list.
    -- This is not required, but it provides a place for a description.
    vim.api.nvim_create_autocmd("User", {
        pattern = "RefineryQuickFixCmdPost",
        command = "execute",
        group = refinery_augroup,
        desc = "No-op autocommand that provides a hook event for refinery.nvim",
    })
```

In another file, after I call `setqflist`, I fire that autocommand:

```lua
local on_exit = function(stream, title, efm)
    -- ...
    vim.fn.setqflist({}, " ", {
        lines = lines,
        title = title,
        efm = efm,
    })

    vim.api.nvim_exec_autocmds(
        "User",
        { pattern = "RefineryQuickFixCmdPost", modeline = false }
    )
end
```

The user can then set autocommands on `RefineryQuickFixCmdPost` as if that were a built-in autocommand:

```lua
-- Open quickfix window after refinery.nvim adds items to it.
create_autocmd("User", {
    pattern = "RefineryQuickFixCmdPost",
    command = ":cwindow",
    group = telemachus_augroup,
})
```

**Comment by hfn92** [comment]
@telemachus hey thanks for helping me out with this :) But I guess this would only work if you have control over when `setqflist` gets called, right? I use plugins that essentially call `vim.fn.setqflist({}, "a", { lines = { ... } })` to feed lines to the qflist. I guess I would need to have them support my custom autocommand for this to work.

**Comment by telemachus** [comment]
@hfn92 Yes, sorry. In my case, I wrote the plugin myself. If it's someone else's plugin, then you need to encourage them to adopt this approach. In that case, they wouldn't be supporting *your* custom autocommand. Instead, they would add their own custom autocommand, and you would use that in your configuration files. For example, see the [autocommands section in paq's documentation](https://github.com/savq/paq-nvim/blob/master/doc/paq-nvim.txt#L339-L355).

tl;dr - you need to lobby the plugins to include an autocommand hook.
